### PR TITLE
Make filtering on mutation types configurable

### PR DESCRIPTION
**Pull request by oplantalech**
# What? Why?
Currently, cBioPortal filters out the following mutation types: `Silent`, `Intron`, `IGR`, `3'UTR`, `5'UTR`, `3'Flank` and `5'Flank`, except the promoter mutations for the TERT gene. We want to make this filtering by mutation type configurable by the user that loads studies in cBioPortal.

Changes proposed in this pull request:
- Add an optional field in the mutation meta file that specifies the mutation types that want to be filtered out. If the field is left empty, no mutation types are filtered. If the field is not written in the meta file, the default filtering done by cBioPortal is applied.

The parameter is called `variant_classification_filter`. The mutation types need to be written separated by a comma and following the [MAF nomenclature for Variant Classification](https://wiki.nci.nih.gov/display/TCGA/Mutation+Annotation+Format+(MAF)+Specification).

[More information about this PR](https://docs.google.com/document/d/1Yu_Nu3Q-qdKFiZGf-qJW_Rc0iHPgQTE4m28UX5vTsg0/edit#heading=h.t10nr9te3qvu).

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java [hunk 1]
```diff
@@ -109,53 +110,57 @@ public boolean acceptMutation(ExtendedMutation mutation) {
...
       }
-
-      // Do not accept IGR Mutations
-      if( safeStringTest( mutation.getMutationType(), "IGR" ) ){
-		  igrRejects++;
-         return false;
+      
+      //Filter by types if specified in the meta file, else filter for the default types
+      if (filteredMutations != null) {
+          Set<String> variantFilteredMutations = new HashSet<String>();
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: please make this global and initialize it (i.e. the loop on lines 130-133) in constructor of this class.

Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: otherwise this initialization code will execute for each mutation

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java [hunk 2]
```diff
@@ -254,6 +235,18 @@ public int getRedactedRejects()
 	{
 		return this.redactedRejects;
 	}
+	
+   public Map<String, Integer> getMapRejections() {
+       return this.countMap;
+   }
+   
+   public void addRejectedVariant(Map<String, Integer> countMap, String mutation) {
+       if (this.countMap.get(mutation) != null) {
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: I thought the `computeIfPresent` made the if / else unnecessary....is this not the case?

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java [hunk 3]
```diff
@@ -254,6 +235,18 @@ public int getRedactedRejects()
 	{
 		return this.redactedRejects;
 	}
+	
+   public Map<String, Integer> getMapRejections() {
+       return this.countMap;
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: please call this `mapRejections` or rename getter to `getCountMap`

File: core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java [hunk 4]
```diff
@@ -275,4 +276,10 @@ public static String loadGenePanelInformation(File file) throws Exception {
         properties.load(new FileInputStream(file));
         return properties.getProperty("gene_panel");
     }
+
+	public static String loadFilteredMutations(File file) throws Exception {
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java]: I would call this getVariantClassificationFilter. Please document here what the method does (gets a setting) and briefly what the purpose of this variant_classification_filter setting is .

File: core/src/main/scripts/importer/cbioportalImporter.py [hunk 5]
```diff
@@ -363,7 +363,7 @@ def main(args):
             print 'Data loading step using: {}\n'.format(args.jar_path)
         
     # process the options
-    jvm_args = "-Dspring.profiles.active=dbcp -cp " + args.jar_path
+    jvm_args = "-Dspring.profiles.active=dbcp -cp " + args.jar_path #-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=8000
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/cbioportalImporter.py]: please remove this

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java (hunk 2 above)
Comment (oplantalech) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: > If the value for the specified key is present and non-null, attempts to compute a new mapping given the key and its current mapped value.
>
> If the function returns null, the mapping is removed. If the function itself throws an (unchecked) exception, the exception is rethrown, and the current mapping is left unchanged.

So it only works if the key is already present.

File: core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java [hunk 6]
```diff
@@ -277,7 +277,16 @@ public static String loadGenePanelInformation(File file) throws Exception {
         return properties.getProperty("gene_panel");
     }
 
-	public static String loadFilteredMutations(File file) throws Exception {
+	/**
+	 * Gets the information of "variant_classification"filter" in the file, if it exists. Otherwise, it
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java]: typo "variant_classification"filter" -> "variant_classification_filter"

File: core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java [hunk 7]
```diff
@@ -277,7 +277,16 @@ public static String loadGenePanelInformation(File file) throws Exception {
...
+	 * returns null. "variant_classification_filter" can be used in the mutation meta file to specify
+	 * which types of mutations want to be filtered.
+	 * 
+	 * @param file
+	 * @return a string with the types of mutations that want to be filtered, comma-separated.
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java]: maybe rephrase "types of mutations that want to be filtered" -> types of mutations that should be filtered

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java (hunk 2 above)
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: try combining it with `computeIfAbsent()`

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java [hunk 8]
```diff
@@ -236,15 +231,15 @@ public int getRedactedRejects()
...
    
-   public void addRejectedVariant(Map<String, Integer> countMap, String mutation) {
-       if (this.countMap.get(mutation) != null) {
-           this.countMap.computeIfPresent(mutation, (k, v) -> v + 1);
+   public void addRejectedVariant(Map<String, Integer> rejectionMap, String mutation) {
+       if (this.rejectionMap.get(mutation) != null) {
+           this.rejectionMap.computeIfPresent(mutation, (k, v) -> v + 1);
        } else {
-           this.countMap.put(mutation, 1);
+           this.rejectionMap.put(mutation, 1);
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: instead of if/else try using `computIfAbsent()`

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java [hunk 9]
```diff
@@ -78,8 +78,11 @@ public ImportExtendedMutationData(File mutationFile, int geneticProfileId, Strin
         this.geneticProfileId = geneticProfileId;
         this.swissprotIsAccession = false;
         this.genePanel = genePanel;
-        this.filteredMutations = filteredMutations;
-
+        for (String filteredMutation : (Arrays.asList(filteredMutations.split(",")))) {
+            filteredMutation = filteredMutation.trim();
+            this.filteredMutations.add(filteredMutation);
+        }
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java]: this can be simpler:
```
this.filteredMutations = new HashSet<String>(Arrays.asList(filteredMutations.split(",")));
```
that should be enough.

Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java]: Actually, I also think this would be better done in `getVariantClassificationFilter` immediately. This makes it clear that this setting is about a set of values.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java [hunk 10]
```diff
@@ -109,53 +110,52 @@ public boolean acceptMutation(ExtendedMutation mutation) {
...
+                   safeStringTest( mutation.getMutationType(), "3'Flank" ) ||
+                   safeStringTest( mutation.getMutationType(), "5'UTR" ) ||
+                   safeStringTest( mutation.getMutationType(), "IGR") ){
+              addRejectedVariant(rejectionMap, mutation.getMutationType());
+              return false;
+          }
+          
+          if( safeStringTest( mutation.getMutationType(), "5'Flank" ) ) { 
+                if (whiteListGenesForPromoterMutations.contains(mutation.getEntrezGeneId())){
+                      mutation.setProteinChange("Promoter");
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: this is important and should apply to all 5'flank mutations

File: core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java [hunk 11]
```diff
@@ -278,17 +281,23 @@ public static String loadGenePanelInformation(File file) throws Exception {
...
 	 * @throws Exception
 	 */
-	public static String getVariantClassificationFilter(File file) throws Exception {
+	public static Set<String> getVariantClassificationFilter(File file) throws Exception {
 	    Properties properties = new TrimmedProperties();
 	    properties.load(new FileInputStream(file));
-	    return properties.getProperty("variant_classification_filter");
+	    String variantClassificationFilter = properties.getProperty("variant_classification_filter");
+	    Set<String> filteredMutations = new HashSet<String>();
+	    for (String mutation : (Arrays.asList(variantClassificationFilter.split(",")))) {
```
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java]: please try 
```
Set<String> filteredMutations = new HashSet<String>(Arrays.asList(filteredMutations.split(",")));
```
instead of this loop

Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java]: or maybe leave it...because of mutation.trim

**Comment by pieterlukasse** [comment]
@oplantalech to finish, can you please setup a unit tests wit the following scenarios:
(prerequisite) setup dummy mutation file with the various filtering cases
1- check that the expected events are filtered out when running in default mode
2- check that the expected events are filtered out when using the filter property

Please also check that the events filtered out in (1) above are the same when running the previous version of the code (before your changes). 
:information_source: see `/cbioportal/core/src/test/java/org/mskcc/cbio/portal/scripts/TestImportExtendedMutationData.java`; there are already some unit tests there.

**Comment by oplantalech** [comment]
@pieterlukasse I only needed to add tests to check for the filtering property. I already added this piece of code which allows to reuse the tests that are already there as having the default filtering:
```
public ImportExtendedMutationData(File mutationFile, int geneticProfileId, String genePanel) {
        this(mutationFile, geneticProfileId, genePanel, null);
    }
```

**Comment by oplantalech** [comment]
@pieterlukasse 
rc:
```
Filtering table:
-----------------
Mutation filter decisions: 87707
Rejects: 23477
Mutation Status 'None' Rejects:  0
Silent or Intron Rejects:  21543
UTR Rejects:  1934
IGR Rejects:  0
LOH or Wild Type Rejects:  0
Empty Annotation Rejects:  0
Missense Germline Rejects:  0
```

PR:
```
Filtering table:
-----------------
Mutation filter decisions: 87707
Rejects: 23477
Mutation Status 'None' Rejects:  0
LOH or Wild Type Rejects:  0
Empty Annotation Rejects:  0
Missense Germline Rejects:  0
5'Flank Rejects: 343
3'Flank Rejects: 288
Intron Rejects: 2875
Silent Rejects: 18668
5'UTR Rejects: 386
3'UTR Rejects: 917
```

`Silent or Intron Rejects` = `Intron Rejects` + `Silent Rejects`
`UTR Rejects` = `5'Flank Rejects` + `3'Flank Rejects` + `5'UTR Rejects` + `3'UTR Rejects`

With the new system if they are no rejections that are not `Mutation Status 'None'`, `LOH or Wild Type`, `Empty Annotation` and `Missense Germline`, the category do not appear in the table. Therefore, there are no `IGR Rejects` in the PR.

**Comment by pieterlukasse** [comment]
Nice, I like the new filtering table better.

**Comment by oplantalech** [comment]
I just squashed all the commits into one, so now the PR is ready to merge.

**Comment by pieterlukasse** [comment]
temporarily changed to "do not merge"...one part for the validation script is missing.

**Comment by jjgao** [comment]
@n1zea144 or @sheridancbio could you please also review it?

**Comment by oplantalech** [comment]
@pieterlukasse @jjgao @n1zea144 @sheridancbio I removed the "do not merge" tag. The PR is ready to be reviewed.

File: core/src/main/scripts/importer/validateData.py [hunk 12]
```diff
@@ -1258,6 +1258,10 @@ def skipValidation(self, data):
         """Test whether the mutation is silent and should be skipped."""
         is_silent = False
         variant_classification = data[self.cols.index('Variant_Classification')]
+        if 'variant_classification_filter' in self.meta_dict:
+            self.SKIP_VARIANT_TYPES = [x.strip()
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: what if the configured types are not valid types? Then the warning from `checkVariantClassification` is not given anymore for applicable cases, right?

File: core/src/test/scripts/unit_tests_validate_data.py [hunk 13]
```diff
@@ -1106,6 +1106,26 @@ def test_alternative_notation_for_intergenic_mutation(self):
...
+        old_variant_types = validateData.MutationsExtendedValidator.SKIP_VARIANT_TYPES
+        validateData.MutationsExtendedValidator.SKIP_VARIANT_TYPES = ["5'Flank", "Frame_Shift_Del", "Frame_Shift_Ins"]
+        record_list = self.validate('mutations/data_mutations_some_silent.maf',
+                                    validateData.MutationsExtendedValidator,
+                                    extra_meta_fields={
+                                            'swissprot_identifier': 'name'})
+        # we expect 6 infos: 4 about filtered mutations, 2 general info messages:
+        self.assertEqual(len(record_list), 6)
+        # First 3 INFO messages should be something like: "Line will not be loaded due to the variant classification filter. Filtered types:"
+        for record in record_list[:3]:
```
Comment (pieterlukasse) [inline core/src/test/scripts/unit_tests_validate_data.py]: 3 or 4?

File: docs/File-Formats.md [hunk 14]
```diff
@@ -499,6 +499,9 @@ The mutation metadata file should contain the following fields:
 8. **data_filename**: your data file
 9. **gene_panel**: optional gene panel stable id
 10. **swissprot_identifier (optional)**: either `accession` or `name`, indicating the type of identifier in the `SWISSPROT` column
+11. **variant_classification_filter (optional)**: mutation types to be filtered out, separated by commas, using the allowed values specified below. By default, cBioPortal filters out the following mutation types: Silent, Intron, IGR, 3'UTR, 5'UTR, 3'Flank and 5'Flank, except the promoter mutations for the TERT gene. If no types want to be filtered out, leave the field empty.
```
Comment (pieterlukasse) [inline docs/File-Formats.md]: If you leave it empty, will no types be filtered out or will the default cBioPortal filtering apply? Please clarify.

File: core/src/test/scripts/unit_tests_validate_data.py [hunk 15]
```diff
@@ -1121,7 +1121,7 @@ def test_customized_variants_skipped(self):
         # we expect 6 infos: 4 about filtered mutations, 2 general info messages:
         self.assertEqual(len(record_list), 6)
         # First 3 INFO messages should be something like: "Line will not be loaded due to the variant classification filter. Filtered types:"
```
Comment (pieterlukasse) [inline core/src/test/scripts/unit_tests_validate_data.py]: 3 or 4?

File: core/src/main/scripts/importer/validateData.py [hunk 16]
```diff
@@ -1258,6 +1258,14 @@ def skipValidation(self, data):
...
         is_silent = False
         variant_classification = data[self.cols.index('Variant_Classification')]
+        if 'variant_classification_filter' in self.meta_dict:
+            self.SKIP_VARIANT_TYPES = [x.strip() 
+                                       for x 
+                                       in self.meta_dict['variant_classification_filter'].split(',')]
+        
+        # check that the values written by the user are valid values, otherwise raise a warning
+        for value in self.SKIP_VARIANT_TYPES:    
+            self.checkVariantClassification(value)
```
Comment (pieterlukasse) [inline core/src/main/scripts/importer/validateData.py]: I would not give a warning here, since these values will be skipped/ filtered. The warning given here is anyway not accurate for this scenario.

**Comment by sandertan** [comment]
@oplantalech Nice PR. Could you add `variant_classification_filter` to `MetaFileTypes.MUTATION` in `cbioportal_common.py`?

I'm getting this warning in my validation report:
`Unrecognized field in meta file	variant_classification_filter`

**Comment by oplantalech** [comment]
@sandertan Done.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java [hunk 17]
```diff
@@ -109,53 +110,55 @@ public boolean acceptMutation(ExtendedMutation mutation) {
...
       
-      if( safeStringTest( mutation.getMutationType(), "5'Flank" ) ) { 
-            if (whiteListGenesForPromoterMutations.contains(mutation.getEntrezGeneId())){
+      //Filter by types if specified in the meta file, else filter for the default types
+      if (filteredMutations != null) {
+          if (filteredMutations.contains(mutation.getMutationType())) {
+              addRejectedVariant(rejectionMap, mutation.getMutationType());
+              return false;
+          } else {
+              if( safeStringTest( mutation.getMutationType(), "5'Flank" ) ) {
```
Comment (oplantalech) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: That's a good question. Indeed, my idea here was to maintain the functionality exactly as it was if no `variant_classification_filter` is specified. That also means skipping the promoter mutations of the TERT gene. However, it might be a good idea to discuss with the rest of the community whether we want to remove this specific check for the TERT promoter mutations.

File: core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java [hunk 18]
```diff
@@ -275,4 +279,29 @@ public static String loadGenePanelInformation(File file) throws Exception {
...
+	    String variantClassificationFilter = properties.getProperty("variant_classification_filter");
+	    if (variantClassificationFilter != null) {
+		    Set<String> filteredMutations = new HashSet<String>();
+		    for (String mutation : (Arrays.asList(variantClassificationFilter.split(",")))) {
+		            mutation = mutation.trim();
+		            filteredMutations.add(mutation);
+		        }
+		    return filteredMutations;
+	    } else {
+	    		return null;
```
Comment (oplantalech) [inline core/src/main/java/org/mskcc/cbio/portal/util/GeneticProfileReader.java]: Thanks for noticing! I've fixed it.

**Comment by oplantalech** [comment]
@jjgao Can you merge this PR? Thanks.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java (hunk 17 above)
Comment (pieterlukasse) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java]: Jsut for the record: I think that  when a variant_classification_filter *is* specified it should be leading and there should be no whitelist. IMO a whitelist would lead to confusion in this case.
